Proposal of macro processing and showcase of new instructions

ochaloup opened the pull request:
@aankor @luciotato after some exploration I decided to spend some time on Rust macro processing and thus proposing this PR with (simplified) macros specialized for Data and Accounts for these particular SDK instructions.

This PR contains as the main change the macro processing and three new SDK instructions (where one of them is just scripted manually without macro usage). I would like to know if you are fine with this approach.
After adjustments and fixes I would be happy if I can merge this and then proceed further with all other instructions in the way suggested here.

(I'm not sure if the PR is not a bit big, I could split it to few more - e.g. some formatting could be strip off etc.)
(Currently there are no tests for SDK instructions but I plan to either add some or work with tests through https://github.com/marinade-finance/liquid-staking-referral-program as @luciotato suggested to me)

ochaloup committed:
Adding instruction add_liquidity

ochaloup committed:
Formatting with cargo fmt

ochaloup committed:
Adding instruction change_authority

ochaloup committed:
No warnings on build for not used imports

ochaloup committed:
SDK instruction macros added

ochaloup committed:
.idea to gitignore

ochaloup committed:
Adding instruction deposit_stake_account

ochaloup commented:
@aankor I fixed the thing about the `ID` attribute to not be taken from context. Now I'm looking about that nesting.

ochaloup commented:
I added a check that does not permit to run the macro with struct of fields that are not all `Pubkey`.

ochaloup committed:
Enhancing add_liquidity to with macro definitions

ochaloup commented:
Update: Finding a way how to work with nested struct. The macro implementation In progress.

ochaloup committed:
Macro processing to work with nested structs

ochaloup commented:
@aankor I changed the macro code to be capable to manage nested structs. I believe there is a space for optimizations thus I look forward your deeper review.

aankor commented on libs/marinade-sdk/src/instructions/add_liquidity.rs:
```diff
@@ -0,0 +1,212 @@
+    pub mint_to: Pubkey,
+    pub system_program: Pubkey,
+    pub token_program: Pubkey,
+    pub test_nested: NestedOneAccounts,
```
Remove temp code

ochaloup replied:
Fixed, thanks.

aankor commented on libs/marinade-sdk-macro/src/lib.rs:
````diff
@@ -0,0 +1,455 @@
+/// ```
+///
+/// ```
+/// use marinade_sdk_macro::MarinadeInstructionData;
````
Looks like duplication of structs

ochaloup replied:
I'm sorry I'm not sure what do you mean by the "duplication of structs"?
The `use` statement needs to be part of the definition otherwise the comment code is not compiled.

ochaloup commented:
@aankor Thank you for the review. I fixed my omission at liquidity instruction and added a question on the second comment.
Can you take a look, please?

ochaloup commented:
Just a small update as I found wrong handling of empty TokenStream in macro. Fixed here.
Still ready for the review.

ochaloup committed:
Macro processing to work with account attributes of ownerid and data struct name

ochaloup commented:
The accounts struct attribute may define what is the name of the data struct that it depends on.

ochaloup commented:
Closing this PR. It's replaced by pr #2.
